### ISSUE TITLE
Don’t parse the pipeline as text when it is directed from an EXE to another EXE or file. Keep the bytes as-is.

### ISSUE THREAD

**Issue by be5invis** [PowerShell/PowerShell#1908]
Currently PowerShell parses STDOUT as string when piping from an EXE, while in some cases it should be preserved as a byte stream, like this scenario:

``` powershell
curl.exe http://whatever/a.png > a.png
```

or

``` powershell
node a.js | gzip -c > out.gz
```

Affected patterns include: `native | native`, `native > file` and (maybe) `cat file | native`.

**Comment by be5invis** [comment]
@vors @lzybkr 
The current `NativeCommandProcessor` breaks:

- LF line endings.
- Non-ASCII text within UTF-8 without BOM header.
- Binary file redirects (like `curl.exe`’s output).
- `>` layouts text into 80 columns by default.

**Comment by ForNeVeR** [comment]
Maybe it'd be better to add another set of pipe operators that wouldn't parse the output as strings?

**Comment by ForNeVeR** [comment]
Or add a cmdlet/operator to call native command and get its raw output (as a byte array / stream?), something like this:

```powershell
$output = ^& curl.exe http://whatever/a.png # $output now is a byte array or stream
$output > C:\Temp\file.png # file.png now is a valid image file

# This should be valid, too:
^& curl.exe http://whatever/a.png > C:\Temp\file.png
```

This opens an opportunity for some additional usage patterns (you can put this raw content to variables, and pipe raw content from native commands to managed cmdlets).

**Comment by ForNeVeR** [comment]
But maybe we could add a special kind of redirection operator (like `2>&1`, `3>&1`, `*>1` we already have), something like this (where `%>1` is a new redirection operator that redirects command "raw output" without processing it as a string):

```powershell
$output = curl.exe http://whatever/a.png %>1
$output > C:\Temp\file.png

# Or even this:
curl.exe http://whatever/a.png %> C:\Temp\file.png # which is just awesome
```

**Comment by be5invis** [comment]
@ForNeVeR My proposal is that:

1. For `native | native`, keep the bytes as-is. This is already purposed by @vors.
2. For `ps | native`, add a set of cmdlets which encodes PS objects into bytes, perhaps `ps | encode-text utf-8 | native`.
3. For `native | ps`, we can use the type system to identify whether a cmdlet accepts “raw input”. For cmdlets like `out-file` or maybe `decode-text`, it will keep the bytes from `native`, and other cmdlets will use the parsed string as its input.

**Comment by ForNeVeR** [comment]
@be5invis okay, it seems like this proposal also supports all the relevant use cases I can imagine.

**Comment by GeeLaw** [comment]
Shouldn't this open up an RFC since this is a breaking change (changes the observed behaviour)?

A workaround for this is to provide a cmdlet that stores the content in a temporary file. A working example is `Use-RawPipeline` in PowerShell Gallery. The current implementation is to store the file, but it could also be streamlined so that the file doesn't have to be stored.

**Comment by jhclark** [comment]
See also #559, where this appears to be actively discussed and worked on by @vors on the PowerShell team.

**Comment by vors** [comment]
Great discussion! Thank you all for the feedback.

I'd like to share my plans about this work:

* In the scope if this issue we will address only `native | native` and `native > file` behavior. Note, that although it could be seen as a breaking change, it would not be the case for text output. The behavior would be preserved. Byte output would be much more reliable without wrapping bytes in PS strings. We agreed with @lzybkr that it's not breaking, hence no RFC process would be applied.

* I don't see the immediate need in enhancing `native | ps` case, since PS is able to consume strings only from the native commands. Although, somebody may want to write function like

```powershell
function foo
{
  param([byte[]]$rawBytes)
}
```

they may archive it with a temp file or some other technique as @GeeLaw pointed out.

* Similarly, `ps | native` case has a well established pattern: when ps objects need to be passed to the native command, we apply implicit `Out-String` and pass everything as a text.
Because PS doesn't use byte streams as a primitive for pipeline, I don't think we should develop special sugar to support it in the language directly. If there is a case, when it needs to be done, similar work-arounds can be used.

We can revisit the last two parts later, but I'd like to set expectations about scope of this issue.

**Comment by be5invis** [comment]
@vors However the current “>” is identical to `out-file`, so you have to add a special version of `out-file` which takes raw bytes. So why don’t you give the ability to everyone?

**Comment by masaeedu** [comment]
@vors The change in #2450 greatly improves the experience, but the design still feels a little awkward and inconsistent. It seems to be based on arbitrary patterns rather than consistent behavior of operators and cmdlets with respect to input arguments of particular .NET types.

As a user I would expect binary operators like `|` to behave consistently given a LHS expression that evaluates to a byte stream (or some appropriate choice of byte stream-ish object), regardless of whether it is produced by invocation of a native executable, piping from a file, invocation of a PowerShell function/script/cmdlet, or .NET FFI. 

Similarly, I would expect the `|` operator to behave consistently given a RHS that "can accept stdin", for some meaning of accepting stdin appropriate to the RHS expression in question. For native executables and files this is just sending the bytes to the correct file descriptor, for PowerShell invocables perhaps it would be `param([Stream]$rawBytes)` as you suggested.

If there is no way to overload `|`'s behavior so that this is not a breaking change, then we should have a different piping operator for raw streams, and cmdlets for converting between byte streams and guessed-encoding-decoded lines (similar to `$` and `~` from @GeeLaw's [`Use-RawPipeline`](https://psguy.me/modules/Use-RawPipeline/) project).

**Comment by vors** [comment]
Oh, https://psguy.me/modules/Use-RawPipeline/ is very interesting, thank you for the link.

There are 2 conversations going on here:
1. just `native | native` (or `native > file`)
2. `native | ps` and `ps | native`

The are highly related and it's true that solving (2) in a general way will buy us (1) automatically.
However the scope of the work for (2) is much broader and includes RFC and what not, while the (1) is a low hanging fruit: it can be done in a non-breaking manner, greatly improve perf for common cases and the changes itself are very modest. Note that to achieve perf parity with bash, we would need (1) in one form or another.

That's why I think that it make sense to separate these two tasks.

**Comment by masaeedu** [comment]
@vors What I mean is that IMHO `native`, `ps` etc. shouldn't be distinct, first class concepts in the first place. It makes the language conceptually simpler if it consists only of expressions that can produce and consume .NET values and operators that can wire such expressions together. PowerShell is after all strongly typed, if not statically typed.

Changing the language so that `(ping.exe).GetType()` is `Stream` (or some similar type with suitable metadata about the process) would be a way to synthesize native commands with the rest of Powershell in a less inconsistent way.

> That's why I think that it make sense to separate these two tasks.

Do you know if there's already an RFC or issue for the latter task?

**Comment by GeeLaw** [comment]
FYI `Use-RawPipeline` has been reworked to allow streamlined experience instead of having to store the content in a file and wait before the previous process ends to perform the next piped process. Its source code is available from https://github.com/GeeLaw/Use-RawPipeline

**Comment by btjwork** [comment]
There seems to be an assumption throughout this that PowerShell's philosophy that "everything is a pipeline" is OK.  However, I think there might be some value in thinking of the use case of "PowerShell as a legacy native command launcher" as distinct.  Would it be possible to allow the redirection operators to have their traditional meaning of redirecting the native command's stdout directly to the raw file instead of piping its output back to PowerShell?  Even if all of the encoding issues are resolved for general piping, giving a command a pipe when it expects a physical file is still a semantic change.

Requiring a user to know that they need to specify special obscure options to say "don't change the output of this command" seems error-prone at best.  I'm arguing that redirecting a simple command directly to a file should be data-preserving by default.

Or, is the philosophy that if someone wants to use non-text native commands that they should just switch back to a traditional cmd window?

**Comment by guhuajun** [comment]
@SteveL-MSFT 

Following command will be broken due to this issue.
docker save microsoft/windowsservercore:ltsc2016 > msft_wsc_ltsc2016.tar

You have to use docker save -o instead.

**Comment by mpawelski** [comment]
I have hard time understanding what #2450 actually fix? Because even though command like `ping.exe github.com | grep.exe Reply` works. The command:
`curl.exe "https://i.redd.it/dntes9fqy3x11.jpg" > test1.jpg`
still works only in cmd/WSL's bash/git bash.

I was trying to use `git show ref:path/to/file.png > file.png` and it looks that it's still not possible to use it in powershell. Are there any serious plans to fix it?

**Comment by GeeLaw** [comment]
@mpawelski

#2450 fixes the problem of requiring upstream native command to finish before piping the output downstream. It does not address PowerShell’s parsing byte stream from native output and reserializing object to byte stream. Please take some time to learn about how object-oriented pipes work and you will learn the problem is really hard to solve consistently. For that purpose, please use a native redirection/piping utility, e.g., Command Prompt, Start-Process, or Use-RawPipeline.

**Comment by be5invis** [comment]
@GeeLaw So you think PS should break simple, native-exe piping, **on purpose??????**

**Comment by megapro17** [comment]
No-One-Gives-A-Shit-Consider

**Comment by potatoqualitee** [comment]
Re-requesting eval due to this being a blocker for Linux users who want to make pwsh their default shell in Linux.

> One of the main reasons why I haven’t swapped over to pwsh over bash is due to the pipelining issues. Pwsh doesn’t work well when you pipe raw bytes between native apps which is a fairly common scenario

**Comment by daxian-dbw** [comment]
Re-assigned to @JamesWTruher, who will work on this in 7.3 development cycle.

**Comment by 89z** [comment]
https://github.com/PowerShell/PowerShell/issues/15301

**Comment by AE1020** [comment]
I am interested to know how many people who want the "as-is" bytes are also in the camp of people who are using LF instead of CR LF on their Windows PowerShell scripts...

I've proposed a mode that guides behavior based on detection of LF vs. CR LF of the script containing the pipe/redirect:

https://github.com/PowerShell/PowerShell/issues/16511

It *may* be (?) that someone's feelings about the importance of "as-is" redirection vs CR LF is effectively captured by their git autocrlf setting.  So this would piggyback on that.

**Comment by UberKluger** [comment]
Perhaps a useful implementation would recognise the historical significance of a `[byte]` stream as it pertains to native applications. Three scenarios would need to be considered:

1) Whenever a stream of *actual* `[byte]`s is passed into a pipeline to a native program (e.g. `Get-Content -AsByteStream somefile | native`) then no `Out-String` conversion is applied, the bytes are streamed *as-is*. It seems unlikely that many native apps would be expecting a sequence of decimal representations of the value of the byte, one per line, and this could easily be created by simply converting the `[byte]` stream into an `[int]` stream (which would then still use `Out-String`). Of course, there would be no practical way for PS to know that a stream consists entirely of `[byte]`s unless the (single) object passed were a `[byte[]]` (which would then be enumerated into a `[byte]` stream). It would be a matter for debate as to whether only this particular case would trigger the special behaviour or any stream with an initial `[byte]` would trigger it but then throw a terminating error if a non-`[byte]` were passed (similar to `Set-Content -AsByteStream` which has the same issue when taking input `[object]`s from the pipeline).

2) For the case where a native program is the pipeline input to a cmdlet (e.g. `native | Sort-Object`), the current behaviour is retained since often (if not mostly) the native program's output will be (typ. ASCII) text. Bytes are converted to unicode characters by whatever mapping is appropriate and then collected into `[string]`s which are passed to the cmdlet. However, similar to the way commandline arguments are parsed into objects but the original text is retained in case a native command is being invoked, each `[string]` object would be wrapped in a `[psobject]` (or just have an added property) which would contain the original `[byte]`s received from the native program (maybe a `[byte[]]` or a `[string]` with the original `[byte]`s collected but not mapped to unicode equivalents). This would be invisible to existing cmdlets (possibly directly accessible if desired via a public property) but a new filter cmdlet (`ConvertTo-ByteStream`  or `cbs` perhaps) would be provided which would restore the original output into a `[byte]` stream which could then be passed into the remaining pipeline. This would also cover the case of saving native output into a file by simply doing `native | ConvertTo-ByteStream | Set-Content -AsByteStream` (the behaviour of `>` would be unaltered to minimise potential breaking changes).

3) For `native | native ( | native ...)`, I would suggest that the basic ethos of the PS pipeline (passing objects) is totally inappropriate as such programs are not suited for (nor even aware of) the PS pipeline. For this case (only), the original `cmd` (and Unix&trade;) behaviour should be restored. Each program would be started in its own process (as currently) with a (Windows/Unix&trade;) anonymous byte stream pipe connecting StdOut of each to StdIn of the next (if any). Any sequence of `native | native (| native ...)` within a larger pipeline would be treated as a single native program with one input stream from the PS pipeline and one output stream back to the PS pipeline as per items 1 and 2 above, e.g.
   ```
   Get-Content -AsByteStream somefile | native | native | native | Sort-Object | more
   ```
   would be treated as
   ```
   Get-Content -AsByteStream somefile | ( one native doing 3 things ) | Sort-Object | more
   ```
   While this might seem to be the breakingest change possible, I suspect that the vast majority of native programs (with input / output suitable for piping) would have been designed around this behaviour (byte stream pipes, the type obtained from kernel APIs). Certainly, anything intended for invocation by `cmd` would expect it. In particular, `find.exe`, `findstr.exe` and `sort.exe` do (and also don't like Unicode). Other programs that might have the ability to process Unicode would either utilise a command line option, a BOM (LE / BE) as the first two bytes read or (less predictable) DBCS heuristics, e.g. expecting a lot of alternating 0 bytes for mostly ASCII characters or maybe a LE/BE-Unicode Space, TAB or new-line within the first 100 byte pairs (I'm looking at you, `Scripting.FileSystemObject`), but they would still read a byte stream (as paired bytes). Further on the "make it work like `cmd`/Unix&trade;", any redirections of file handles ***within*** a native pipeline grouping (native | native | ...) would need to operate on the actual process handles as there would be no "PS pipeline" within the grouping, e.g.
   ```
   Get-Content -AsByteStream infile | native1 2> native1.err | native2 2> native2.err | ConvertTo-ByteStream | Set-Content -AsByteStream outfile
   ```
   Here, native1.err and native2.err are connected directly to file handle 2 of native1 and native2, respectively. Alternately,
   ```
   Get-Content -AsByteStream infile | native1 2>&1 | native2 2>&1 | ConvertTo-ByteStream | Set-Content -AsByteStream outfile
   ```
   In this case, streams are merged with file handle 2 of both native1 and native2 being duplicated from each's file handle 1 (their respective anonymous output pipes). Whether (on Unix&trade; like systems) being able to open/redirect/merge file handles other than 2 (&agrave; la sh(1) and derivatives) would be desirable (feasible?) *within* a native pipeline grouping is less clear. There would be few, if any, native programs (on any system) that expected anything beyond StdIn, StdOut and StdErr.

I believe that the preceding could go a long way toward resolving the apparent discontinuity between (legacy) program simple byte oriented CLI I/O and the more powerful but more complicated object oriented PS pipeline. While some potentially breaking changes are involved, I suspect these would mostly affect strategies (kludges?) used to work around the incompatibility issues addressed here. Further, the changes proposed allow original (incompatible) behaviour to be maintained at the PS script level. For native output, no change will occur unless the new `ConvertTo-ByteStream` cmdlet is used. For input, simply changing a `[byte]` stream to an `[int]` stream (of the same values) will restore previous behaviour. For native to native pipes, interposing an explicit `Out-String` between each native in the pipeline should restore previous behaviour by passing the data through PS, with the associated conversions (not 100% sure about this one).

**Comment by huettenhain** [comment]
One way to avoid any magic would be that PowerShell inspects a list of exceptions that the user can configure themselves. It could be as simple as a JSON file stored at a specific location in your user profile:
```json
{
    "exceptions": [
        {
            "path": "C:\\Python39\\python.exe",
            "stdin": "native",
            "stdout": "native",
            "stderr": "native"
        },
    ]
}
```
Obviously, the exact format and location of the configuration isn't so relevant, but this would make it an easy to configure opt-in feature that does not disrupt the way PowerShell works by default.

**Comment by iSazonov** [comment]
> One way to avoid any magic would be that PowerShell inspects a list of exceptions 

General proposal is in #13428

**Comment by hez2010** [comment]
Any updates? Seems that without this we cannot use gzip to compress data stream:

```pwsh
cat -AsByteStream a.js | gzip > a.gz
```

The data stream will be corrupted by the `>`.

**Comment by vexx32** [comment]
@JamesWTruher is this still on your radar?

**Comment by 89z** [comment]
I posted a question about this almost a year ago:

https://stackoverflow.com/questions/67244665/out-file-and-carriage-returns

and I just got bit by this again, today, only to see that this issue is STILL open. Can we PLEASE get some movement on this? Its been literally several years, and this is a constant annoyance. Any time I have to redirect output with PowerShell, I instead have to rewrite my program so that the program creates the output file internally, because PowerShell refuses to just pass the input on as is.

**Comment by 89z** [comment]
Since I am sick and tired of waiting on this, I found another option. NuShell can do the following command, without adding any carriage returns:

~~~
curl www.iana.org/_img/bookmark_icon.ico | save out.ico
~~~

https://github.com/nushell/nushell

**Comment by hez2010** [comment]
Can we expect this being committed in 7.3? Whenever I want to write bytes to native executables via pipeline, I have to launch a cmd or bash shell to achieve this, which makes PowerShell really hard to use to process binary data.

**Comment by lewis-yeung** [comment]
This has annoyed me for a long time. 💢 I believe there are more PS users feeling confused about it. However it's so frustrating that PR #15861 was closed. It's 2022 now, and unfortunately we still cannot get rid of the object-based stream redirection for native executables. 😭

**Comment by mitchcapper** [comment]
lewis-yeung,  it can be quite the pain. The workaround of  https://github.com/GeeLaw/PowerShellThingies/tree/master/modules/Use-RawPipeline requires rewriting your commands but can deliver performance for many situations I have found.

**Comment by lewis-yeung** [comment]
@mitchcapper Thank you. I noticed this amazing work by [GeeLaw](https://github.com/GeeLaw). Hope that a built-in implementation can be applied to the syntax (`>`/`>>` operator) in future PowerShell releases. 🙏

**Comment by AE1020** [comment]
> Hope that a built-in implementation can be applied to the syntax (>/>> operator) in future PowerShell releases.

(+1)  I don't know if it's true for everyone, but if it just required using a different operator to get the effect--that would be fine with me!

e.g. My concern isn't so much "the specific | and > characters don't have the right behavior" as it is *"can't seem to to do it at all"*.

Everything in PowerShell has (to me) a very foreign syntax, so one more adjustment would not be a problem.

**Comment by vors** [comment]
> However it's so frustrating that PR https://github.com/PowerShell/PowerShell/pull/15861 was closed.

Thank you for sharing this PR!
I made a build to see how well it works and it hit a problem for me immidiatlly:

```
PS /Users/vors/src/PowerShell> cat ./README.md | grep CODE
ResourceUnavailable: Program 'cat' failed to run: StandardError has not been redirected.At line:1 char:1
+ cat ./README.md | grep CODE
+ ~~~~~~~~~~~~~~~.
```

Maybe this is because I'm using MacOs, not sure.

**Comment by SeeminglyScience** [comment]
I got a proof of concept of this working a few days ago. There's *a lot* more left to be done, it still needs some abstraction, error handling, redirections in general, and a lot of testing. You would be correct to call this a bug fix, but from a planning and expectations standpoint (i.e. the amount of time it'll take me to finish) it's more in the realm of a feature.

That said, so far so good!

![image](https://user-images.githubusercontent.com/24977523/171234202-f7df747b-82ea-446e-9340-69f27796a8c6.png)

**Comment by btjwork** [comment]
Nope. Excellent question though.

My reasoning (in case others can spot a flaw): In .NET, char and byte are unambiguously distinct types (not even described on the same page in Microsoft's reference). So, if I even managed to create a "raw bytes" variable in the first place, I could assume I got that as the output from a previous native command and/or by reading a file in "raw" mode. In both of those cases, I'd want the semantics to be the same as if I had redirected/piped that original source directly into my native command in a single command line. Sort of a mathematical "transitive property" I suppose.

Even in my past C# programming, if I ever resorted to creating some sort of "collection of byte", the reason was always to indicate "don't touch/modify this data in any way".

If someone were to even present a hypothetical case where they would want it processed, I'd lean towards saying the user should explicitly pipe it through some sort of "stringify" PowerShell command such as Select-Object, since I would expect that to be such an unusual semantic.

**Comment by CleverCoder** [comment]
Wow... just wow. I found another scenario that bit me for a few hours tonight. I was trying to compare the AWS "fingerprint" for a private key I had to determine what key went with what file. Long story short, due to this unexpected behavior the following commands produced different results between Git bash and PS. (It's more an issue now that I'm trying to embrace Powershell more, and it's becoming the default terminal in many IDE's):

Incorrect value using Powershell pipe:
```
PS D:\dev\AWS Keys> openssl pkcs8 -in .\private.pem -nocrypt -topk8 -outform DER | openssl sha1 -c
(stdin)= 8a:cb:c5:84:4d:a6:a3:5f:ed:03:67:f6:f6:88:a0:bd:02:8e:b5:dc
```

Correct value using Git bash:
```
$ openssl pkcs8 -in .\private.pem -nocrypt -topk8 -outform DER | openssl sha1 -c
(stdin)= 89:e3:a5:d2:84:33:ba:ac:64:55:38:e5:5b:52:1f:9e:42:bf:e5:c2

```
Any seasoned programmer would expect the output stream of bytes to flow between processes. This has me baffled, but I'm glad I've been made aware of this oddity!

**Comment by aress31** [comment]
Any progress or fix on the issue, it is about time to give us the option to pipe bytestream...

**Comment by SeeminglyScience** [comment]
Yep plenty of progress, just a large change with a lot of design considerations.

Speaking of, anyone know of a command that emits binary data to std**err**? My current thinking is to just not touch stderr and leave it to it's current string reading ways.

And in the same vein, I'm thinking that redirecting stderr to stdout (e.g. `2>&1`) should fall back to current string reading behavior for both stdout and stderr. Can anyone think of scenarios where that would not be desirable?

**Comment by daxian-dbw** [comment]
@SeeminglyScience I think it's reasonable to assume stderr always emits text. Can you maybe create a draft PR with your changes so far?

**Comment by AE1020** [comment]
> Speaking of, anyone know of a command that emits binary data to stderr?

My issue would not be emitting binary data, but wanting to preserve the sense of CR/LF line endings.

If the program writing to stderr did not put CR in its output, I don't want to see it in the receiving programs or if written to a file.

**Comment by mpawelski** [comment]
> Speaking of, anyone know of a command that emits binary data to std**err**? My current thinking is to just not touch stderr and leave it to it's current string reading ways.
> 
> And in the same vein, I'm thinking that redirecting stderr to stdout (e.g. `2>&1`) should fall back to current string reading behavior for both stdout and stderr. Can anyone think of scenarios where that would not be desirable?

IMO it's not that critical to avoid current "string reading behavior" in this case because when you use `2>&1` you most likely deal with text output, not binary. 

Though I agree with @AE1020 that it's not just about binary data, even when program output text I want to pipe it or redirect "as is" without any changes by Powershell. 

I think this would be unnecessary exception to "we don't parse native executables output as text in pipelines and redirections" feature. It would be another Powershell gotcha, where it behaves slightly different is some corner case.

**Comment by btjgit** [comment]
I agree with the recent comments. If I write a command line that consists only of native executables, all piping and redirection should be as-is, no parsing of anything as text. If a command is ending lines with newlines only on Windows, that should be preserved when I redirect that output to a file or another executable.

I know this goes against the historical architecture of PowerShell, but my expectation would be that parsing things as text would be only done as a "last resort" if/when the output of a binary executable is being _read_ by a cmdlet, rather than thinking of it as something to decide when the output is _written_.

And, the idea of merging stderr into stdout and having that "taint" how stdout behaves sounds like a particularly confusing idea.

**Comment by SeeminglyScience** [comment]
Draft PR is up #17857

**Comment by hut8** [comment]
One thing that I haven't seen mentioned here is perhaps introducing a new operator. This would be a breaking change simply because the `|` and `>` and other redirection operators change meaning, right?

The functionality in https://github.com/GeeLaw/PowerShellThingies/tree/master/modules/Use-RawPipeline is something that should be part of absolutely any shell. As others have mentioned, the main drawback is that it requires `run native.exe | ...` instead of just `native.exe`. Today I thought, "why not just define a new infix operator myself to smooth over the usage of that module?" I am no PowerShell expert but I believe you can only define a prefix operator, not infix or postfix.

So if you just want a "raw pipe", what about keeping the `|` operator as-is and and calling it something like `|!` ?

Aside from the parsing issues, my other main issue is memory. If I want to do `program | tee big.txt` in bash, practically no memory is used. In powershell, `program | tee-object big.txt` will crash the system with an OOM if the output is big enough. This sort of thing totally prevents me from using powershell as my daily driver.

**Comment by mitchcapper** [comment]
One thing of note here is that I believe native-to-native pipe operators could be updated without really any breaking changes (minus the feed rate).  I think the primary breaking changes are when powershell commands are fed the output (if one was to update `|`).

**Comment by btjgit** [comment]
> One thing of note here is that I believe native-to-native pipe operators could be updated without really any breaking changes (minus the feed rate). I think the primary breaking changes are when powershell commands are fed the output (if one was to update `|`).

Yes, I'm getting a little lost as to why new responses keep popping up implying there is some sort of fundamental problem with fixing this. As has been hashed out for six years now, the goal is to limit the scope of this change to only affect output from binary executables going directly to other binary executable or directly to files, while not affecting any behavior where Cmdlets are involved.

We shouldn't need a new operator or some kind of special syntax to allow binary (some might say "real") commands to behave as their authors expected. It seems the challenge is the implementation and historical PowerShell design more than the objective. Let's not lose track of the objective due to implementation detail.

PowerShell has "Shell" in the name, yet it seems that it treats launching native commands as a secondary use case compared to being a programming language. And Microsoft installs it on Windows as a replacement for cmd.exe in the start menu, not as an optional alternative for advanced users. A user who doesn't know or care about Cmdlets shouldn't be forced to learn a whole new way of writing command lines to avoid subtle data corruption. We're not talking about breaking important functionality, we're talking about a surgical bug fix.

**Comment by ravindUwU** [comment]
![sleepy cat with the PowerShell logo on its little head, captioned "I WOULD BE UNSTOPPABLE" "IF NOT FOR THE BYTE STREAMS"](https://pbs.twimg.com/media/Ffaa_IcagAEy9qj?format=jpg&name=900x900)

**Comment by Shayan-To** [comment]
> We shouldn't need a new operator or some kind of special syntax ...

If PowerShell is ever to introduce a way to get the binary output stream of a binary command (or pipe it into a Cmdlet), through, say, a `Stream` object, then maybe a new operator would be needed, and that might complicate things here as well.

**Comment by btjgit** [comment]
> > We shouldn't need a new operator or some kind of special syntax ...
> 
> If PowerShell is ever to introduce a way to get the binary output stream of a binary command (or pipe it into a Cmdlet), through, say, a `Stream` object, then maybe a new operator would be needed, and that might complicate things here as well.

@Shayan-To, you left out the rest of my paragraph, which I believe encapsulates my counter-argument:

>...It seems the challenge is the implementation and historical PowerShell design more than the objective. Let's not lose track of the objective due to implementation detail.

Let's elevate this to user experience, and put aside the implementation challenges for the moment...

My crude analogy: most GUI systems have a clipboard mechanism for transferring data from one application to another.

One approach would be to say only UTF-16 text is supported, and would be captured via a Ctrl+C keystroke. If the source data wasn't really text, then the data would be "transformed" into that format, even if it resulted in a loss/change of information. The end user would get _something_, but it might not be what they expect. The burden of dealing with that _could_ be put on the user, and (for example) we could expect them to learn that if they want to copy HTML and paste it to an HTML-aware application and preserve the mark-up, they should know to instead know to use Alt+Ctrl+C to do the copy, and so on for each possible data format. That would be roughly equivalent to creating an assortment of new pipe operators in PowerShell.

However, another approach would be something similar to what Windows has done with its clipboard: if the sender has a preferred/native non-text format, and the receiver has indicated that it can accept that format, then a direct transfer is done in that format. And, the user can just use Ctrl+C (or the | operator) in all cases, and in most cases will get the expected result.

(I'm intentionally leaving out the implementation detail of how the clipboard accomplishes this - it's a user experience example, not an implementation example, and a different implementation would likely be more appropriate for PowerShell.)

So, my understanding of this baby-step proposal is:
* If both the sender and receiver are binary executables, use a byte stream (or equivalent).
* If the sender is binary and the receiver is a file, then use whatever the sender prefers (i.e., connect its output directly to the file, or equivalent).
* Otherwise, continue to use the current "powerful" PowerShell behavior.
* Most importantly, don't force the _casual_ PowerShell user to manually make that choice by choosing different operators, options, or syntax when an as-is transfer is possible. Transforming/inferring the data type should be a last resort rather than a required step.

To be clear: I'm NOT proposing that a full general anything-to-anything system needs to be designed to take this first tiny step (a general solution might not be needed in the end). That could be deferred until if/when something came up in the future, such as the use case you describe. This baby step is just to fix an existing bug, which is that data output from traditional executables is being modified in certain well-defined cases where it doesn't need to be, and **that is an unfortunate impediment to PowerShell adoption.** Let's make PowerShell less of an annoyance so it can be unstoppable. (Poor kitten! 😉)

**Comment by SeeminglyScience** [comment]
There's some great discussion happening here about `native | script` pipes but it may be better suited in a new issue. This thread and the associated PR are explicitly about the `native | native` scenario. This issue will likely be closed when that PR is merged and I don't want to see the discussion lost in that.

**Comment by SeidChr** [comment]
so i came by this issue, and wondered how it is not possible to send bytes to the input stream of another command.
while its not possible via the pipeline (yet), there are ways:

```powershell
function Send-Bytes {
    param(
        [string] $To, 
        [string] $Arguments
    )
 
    begin {
        $startInfo = [System.Diagnostics.ProcessStartInfo]::new()
        $startInfo.RedirectStandardInput = $true
        $startInfo.FileName = $To
        $startInfo.Arguments = $Arguments
 
        $process = [System.Diagnostics.Process]::Start($startInfo)
        $inputStream = $process.StandardInput.BaseStream
    }
 
    process {
        $inputStream.WriteByte([byte]$_)
    }
 
    end {
        $inputStream.Flush()
        $inputStream.Close()
        
        $process.WaitForExit()
    }
}
```

enables you to do this:
`[byte]0xf4 | Sent-Bytes -To xxd`

unfortunately i wasn't really able to capture the standard output properly, to pipe it back into powershell (.ReadByte() never returns), but maybe someone more capable wants to try? 😃

**Comment by huettenhain** [comment]
Hey @SeidChr, workarounds like this do exist, see for example [Use-RawPipeline](https://www.powershellgallery.com/packages/Use-RawPipeline/1.3). Unfortunately, I believe that many of the people supporting this issue are not looking for a _workaround_. Speaking for myself, I can assure you that no workaround would be sufficient for me to consider using PowerShell as my daily driver. Literally the only way it can work for me is if I can pipe byte streams from one application to another with a single pipe operator like, honestly, in _any other shell_.

**Comment by aetonsi** [comment]
Fixing pwsh piping/redirecting is **crucially needed**... I cannot use pwsh professionally if i can't rely on basic redirection or piping of data...
Even using workarounds like ``cmd.exe /c myapp `| anotherapp`` has problems due to [buffering and whatnot](https://github.com/PowerShell/PowerShell/issues/19036)... Using third party modules like `use-rawpipeline` also has its problems. And they all require rewriting every invocation that pipes or redirects data, in all of your scripts

**Comment by potatoqualitee** [comment]
Indeed, it's the one thing stopping many Linux users from pwsh adoption as a primary shell. Has there been any progress on this front, pwsh team?

**Comment by SeeminglyScience** [comment]
> Indeed, it's the one thing stopping many Linux users from pwsh adoption as a primary shell. Has there been any progress on this front, pwsh team?

@potatoqualitee yeah, my PR is ready for review now (#17857). It is a large change in a sensitive area of the code base, so don't expect a speedy review, but it's comin'.

**Comment by fjh1997** [comment]
This is my trick
```powershell
[console]::outputencoding=[console]::inputencoding=[System.Text.Encoding]::GetEncoding(1252)
$result=python zipbomb --mode=quoted_overlap --num-files=250 --compressed-size=21179
[IO.File]::WriteAllText("$pwd\result.zip",$result,[System.Text.Encoding]::GetEncoding(1252))
```

**Comment by potatoqualitee** [comment]
Great news, @SeeminglyScience !! Thank you, looking very forward to it.

**Comment by SeeminglyScience** [comment]
Forgot to mark the PR as resolving this issue! The PR was merged on 04/27, should be in the next preview 🎉 Closing

**Comment by potatoqualitee** [comment]
HECK YES! 🚀  Such great news for the non-Windows communities. Thanks so much

**Comment by 4cq2** [comment]
to anyone interested, this is the commit:

https://github.com/PowerShell/PowerShell/commit/2424ad83aa4d44fe9e8f507485744e00c66cde58

personally I will be waiting until the next proper release (not preview) but regardless this is great news - thanks (I guess 7 years late is better than never 😀)

**Comment by mitchcapper** [comment]
wooooooooooo this works: `zstd --decompress Digest-BLAKE-0.05.tar.zst --stdout | gzip -9 | gzip -dc | tar -axvf -` 
```
.\\Digest-BLAKE-0.05/
.\\Digest-BLAKE-0.05/BLAKE.xs
.\\Digest-BLAKE-0.05/Changes
.\\Digest-BLAKE-0.05/ex/
.\\Digest-BLAKE-0.05/ex/benchmark.pl
.\\Digest-BLAKE-0.05/lib/
...
```

For me though, with this feature enabled, you can't actually redirect stdout to a file, it always seems to go to the console with native commands followup bug filed though #19836

**Comment by dt-flo** [comment]
I know this thread is rather old, but I have some issues with the current native to native implementation.
1. For some reason, it is very slow compared to the command processor (cmd.exe) pipeline. No matter what I try, this reaches 45 MB/s for me in the peak when reading from one drive and writing the exact same data to another drive (it doesnt only affect that, reading and writing dummy data was just to make sure the CPU is not the bottleneck). The command processor does it at full capacity of my drives, it gets the data through at 180 MB/s (the max my HDDs support)
2. Executables to receive via stdin cant be expressed with a full path or am I missing something? At the beginning of the pipeline, the executable may be written with the full path without any issues. But the second executable (or later ones) can only be called if the exe file is either in the current working directory or in the path variable. I have not found a way to chain multiple executables together, all with their full path to the exe file.

Are there known issues with the speed? How can I chain multiple native commands, each with its full path?

**Comment by astiob** [comment]
I don’t know about the speed, but I can call full paths just fine, so you may need to describe your issue in more detail:

```
> echo.exe yo | D:\msys64\usr\bin\cat.exe
yo
> echo.exe yo | & D:\msys64\usr\bin\cat.exe
yo
```

**Comment by dt-flo** [comment]
> I don’t know about the speed, but I can call full paths just fine, so you may need to describe your issue in more detail:
> 
> ```
> > echo.exe yo | D:\msys64\usr\bin\cat.exe
> yo
> > echo.exe yo | & D:\msys64\usr\bin\cat.exe
> yo
> ```

You're right, it works. Sigh, I had the parameters in quotation marks.

Regarding speed, I did the following
```
fsutil.exe file createnew test_10gb 10737418240 //Create a test 10 GB file
C:\...\Cygwin64\bin\cat.exe .\test_10gb | & "C:\...\Cygwin64\bin\split.exe" -b500MB - F:\Users\...\tmp\split_test.
```
F is a HDD drive here. The exact same command (except for the ampersand) maxes out my HDDs when I run it in cmd.exe.

**Comment by mitchcapper** [comment]
What version of powershell? What version of windows?   I would do something more straightforward like:
Here is older powershell without proper pipelining
```
dd if=dotnet-sdk-8.0.412-win-x64.exe bs=1MB | dd of=NUL bs=1MB
224+1 records in
224+1 records out
224233512 bytes (224 MB, 214 MiB) copied, 16.5476 s, 13.6 MB/s
```
with pipeline correct:
```
dd if=dotnet-sdk-8.0.412-win-x64.exe bs=1MB | dd of=NUL bs=1MB
224+1 records in
224+1 records out
224233512 bytes (224 MB, 214 MiB) copied, 0.352218 s, 637 MB/s
```

**Comment by dt-flo** [comment]
```
PS >dd if=dotnet-sdk-8.0.414-win-x64.exe bs=1MB | dd of=NUL bs=1MB
220+1 records in
220+1 records out
220218536 bytes (220 MB, 210 MiB) copied, 2.9428 s, 74.8 MB/s
0+38695 records in
0+38695 records out
220218536 bytes (220 MB, 210 MiB) copied, 2.94187 s, 74.9 MB/s
PS >
```
in
```
PS >$PSVersionTable | fl

Name  : PSVersion
Value : 7.5.3

Name  : PSEdition
Value : Core

Name  : GitCommitId
Value : 7.5.3

Name  : OS
Value : Microsoft Windows 10.0.26100

Name  : Platform
Value : Win32NT

Name  : PSCompatibleVersions
Value : {1.0, 2.0, 3.0, 4.0…}

Name  : PSRemotingProtocolVersion
Value : 2.3

Name  : SerializationVersion
Value : 1.1.0.1

Name  : WSManStackVersion
Value : 3.0
```
More specifically, I'm on 26100.6584.
In cmd.exe, I get
```
dd if=dotnet-sdk-8.0.414-win-x64.exe bs=1MB | dd of=NUL bs=1MB
220+1 records in
220+1 records out
220218536 bytes (220 MB, 210 MiB) copied, 1.4477 s, 152 MB/s
0+24811 records in
0+24811 records out
220218536 bytes (220 MB, 210 MiB) copied, 1.48674 s, 148 MB/s
```

**Comment by mitchcapper** [comment]
At those speeds I don't think your issue is related to this, looks like you  are likely maxing your drive read speed there.

**Comment by astiob** [comment]
> you are likely maxing your drive read speed there.

They’re getting a 2× speed in cmd.exe on that same drive.

@dt-flo Do make sure it’s not just due to disk caching, though. (That is, if you run the command first in PS and then in cmd.exe, some of the data may still be cached in memory and not actually be read from disk during the second test at all.)

**Comment by dt-flo** [comment]
I tried running it in PWSH again today:
```
dd if=dotnet-sdk-9.0.305-win-x64.exe bs=1MB | dd of=dotnet_out bs=1MB
226+1 records in
226+1 records out
226741024 bytes (227 MB, 216 MiB) copied, 2.60769 s, 87.0 MB/s
0+51725 records in
0+51725 records out
226741024 bytes (227 MB, 216 MiB) copied, 2.65083 s, 85.5 MB/s
```
Switched to cmd.exe
```
dd if=dotnet-sdk-9.0.305-win-x64.exe bs=1MB | dd of=dotnet_out bs=1MB
226+1 records in
226+1 records out
226741024 bytes (227 MB, 216 MiB) copied, 1.14248 s, 198 MB/s
0+27754 records in
0+27754 records out
226741024 bytes (227 MB, 216 MiB) copied, 1.16856 s, 194 MB/s
```
Switched back to PWSH and tried twice:
```
dd if=dotnet-sdk-9.0.305-win-x64.exe bs=1MB | dd of=dotnet_out bs=1MB
226+1 records in
226+1 records out
226741024 bytes (227 MB, 216 MiB) copied, 2.3447 s, 96.7 MB/s
0+52359 records in
0+52359 records out
226741024 bytes (227 MB, 216 MiB) copied, 2.37055 s, 95.6 MB/s
```
```
dd if=dotnet-sdk-9.0.305-win-x64.exe bs=1MB | dd of=dotnet_out bs=1MB
226+1 records in
226+1 records out
226741024 bytes (227 MB, 216 MiB) copied, 2.36039 s, 96.1 MB/s
0+52114 records in
0+52114 records out
226741024 bytes (227 MB, 216 MiB) copied, 2.387 s, 95.0 MB/s
```
That was in Windows with Cygwin. I thought in order to exclude a possible influence from my AV Bitdefender (even if it's "turned off"), I tried it in Linux on the same drive.

The result is as follows:
Bash:
```
flo@flo-VMware7-1:/tmp$ dd if=dotnet-sdk-9.0.305-win-x64.exe bs=1MB | dd of=dotnetout bs=1MB
226+1 records in
226+1 records out
226741024 bytes (227 MB, 216 MiB) copied, 0.988518 s, 229 MB/s
0+3588 records in
0+3588 records out
226741024 bytes (227 MB, 216 MiB) copied, 0.98907 s, 229 MB/s
```
PowerShell 7:
```
flo@flo-VMware7-1:/tmp$ pwsh
PowerShell 7.5.3
PS /tmp> dd if=dotnet-sdk-9.0.305-win-x64.exe bs=1MB | dd of=dotnetout bs=1MB
226+1 records in
226+1 records out
226741024 bytes (227 MB, 216 MiB) copied, 1.68791 s, 134 MB/s
0+106333 records in
0+106333 records out
226741024 bytes (227 MB, 216 MiB) copied, 1.93509 s, 117 MB/s
PS /tmp>
```
Even if it was in disk cache, PowerShell should be as quick as bash this time because I ran it in PWSH after I ran it in Bash, right?

Edit: I should have said that I tried it on an SSD today, that's why it's faster than yesterday.

**Comment by dt-flo** [comment]
I have made another test on another device (same Windows build as the first test machine). This has a built-in NVMe SSD. I have installed a fresh debian 12 with PowerShell 7.5.3 in Hyper-V. The results were the following:

<img width="1411" height="823" alt="Image" src="https://github.com/user-attachments/assets/deeaea52-dc5f-4fda-998f-bbd91a85b71a" />

**Comment by dt-flo** [comment]
I have now tested this on a third device (this device runs Ubuntu 24 LTS as main operating system, so no virtualization) and again, PowerShell is significantly slower than Bash.
Can anyone confirm that PowerShell 7.5 is way slower compared to any other shell (e.g. Bash on Linux, cmd.exe on Windows) on their device(s) as well?

Should I open a new issue?

Because this actually prevents me from making PowerShell my only (used) shell on Linux systems and using PowerShell to chain multiple executable files.

**Comment by hez2010** [comment]
> Should I open a new issue?

You should definitely open a new issue because it is unrelated to this closed one.

**Comment by GeeLaw** [comment]
Re @hez2010 

> Any updates? Seems that without this we cannot use gzip to compress data stream:
> 
> ```
> cat -AsByteStream a.js | gzip > a.gz
> ```
>
> The data stream will be corrupted by the `>`.
> 
> This command works on every shell except PowerShell.

To make it work generally would bring insufferable performance loss (sometimes even non-halting). In addition, using `Get-Content` to write bytes into the pipeline is very expensive. It's like Java's `ArrayList<Byte>`.

The natural (wrong) consideration for `cmdlet | native` is as follows:

- if all objects written by `cmdlet` to the output stream are `byte` objects, then pipe them unchanged to `native`;
- otherwise, pipe all objects from `cmdlet` to `out-string`, then convert to bytes using some encoding, then pipe the converted bytes to `native`.

This, as a starter, would require `cmdlet` to run to completion before data can be piped to `native`, which violates the intention of pipes. Also, consider `cmdlet` and `native` communicating via a file (that is initially empty):

- `cmdlet` writes 1 byte of `0x41` (ASCII `A`) to output stream, then read the file to see if it begins with byte `0x01`. If so, it writes string `"Hello"`. Otherwise, it writes another byte `0x41`.
- `native` reads 1 byte from `stdin`. If the byte is `0x41`, then it writes `0x00` to the file. If the byte is `0x36` (ASCII `6`), it writes `0x01` to the file.

The execution would never be correct, and a usual implementation does not halt. Whether or not binary pipe is used, must be determined before running the command.

---

Currently the operator `<` is reserved for future use. When it's implemented, it should provide the functionality of redirecting a single file as `stdin`. To obtain concatenated files, I think there needs to be some overhaul on pipe handling design.

**Comment by SeeminglyScience** [comment]
> To make it work generally would bring insufferable performance loss (sometimes even non-halting). In addition, using `Get-Content` to write bytes into the pipeline is very expensive. It's like Java's `ArrayList<Byte>`.

It does work now, piping `byte` or `byte[]` to a native command will send the bytes as is. This was more about enabling scenarios than performance, if perf is the goal then using .NET APIs instead would be recommended.

> This, as a starter, would require `cmdlet` to run to completion before data can be piped to `native`, which violates the intention of pipes. Also, consider `cmdlet` and `native` communicating via a file (that is initially empty):

Currently you could pipe each byte one at a time if you wanted, but that would likely end up as worse performance. Note that piping to a native command is already not halting due to implementation detail.

**Comment by GeeLaw** [comment]
Gotcha and I tried the current behavior and it's a gotcha for me (those are two different kinds of gotcha's lol). The previous example doesn't stand because it assumes the piping mechanism is either all stringification or all bytes (not the case) --- I assumed piping of byte array only applied if the upstream expression has a special form!

The actual behavior, which I think should be documented somewhere:

> If the upstream is non-native and the downstream is native, then each object `o` in the upstream output is looked at individually, if `!ReferenceEquals(o, null) && o.GetType() == typeof(byte)` (i.e., it must be of exactly type `[byte]`) then it's written to the downstream as the byte, otherwise it goes through the stringification procedure.

Some examples that might be helpful in documentation:

```powershell
# Previous meaning: stringify the bytes and send to native
# New      meaning: pipe a.txt into native
get-content -asbytestream a.txt | native

# To reinstate previous meaning, use
get-content -asbytestream a.txt -readcount 0 | native
# or
get-content -asbytestream a.txt | foreach { [int]$_ } | native
```
